--- oidentd-1.9.9/src/Makefile.am~	Thu Aug 16 21:22:33 2001
+++ oidentd-1.9.9/src/Makefile.am	Mon Sep 10 10:32:21 2001
@@ -3,9 +3,6 @@
 SUBDIRS = missing
 sbin_PROGRAMS = oidentd
 
-LIBS+=-lmissing
-LDFLAGS+= -Lmissing
-
 DISTCLEANFILES = \
 	stamp_parse \
 	oidentd_cfg_scan.c \
@@ -24,6 +21,8 @@
 	oidentd_cfg_parse.y \
 	os.c
 
+oidentd_LDADD = $(top_builddir)/src/missing/libmissing.a
+
 EXTRA_DIST = \
 	oidentd_scan.c \
 	oidentd_cfg_parse.c \
