diff -urN oidentd-2.0.0/src/oidentd_masq.c oidentd-2.0.0.my/src/oidentd_masq.c
--- oidentd-2.0.0/src/oidentd_masq.c	Sat Sep 29 21:46:20 2001
+++ oidentd-2.0.0.my/src/oidentd_masq.c	Tue Oct  2 11:15:40 2001
@@ -124,9 +124,11 @@
 
 		p = strtok(buf, " \t");
 		if (p == NULL) {
-			debug("[%s:%d] Missing address parameter", MASQ_MAP, line_num);
-			goto failure;
+			continue;
 		}
+		
+		if (*p == '#')
+			continue;
 
 		temp = strchr(p, '/');
 		if (temp != NULL)
